### PR TITLE
Laboratory work #4, Anastasia Logunova - 22FPL1

### PR DESCRIPTION
**Pull request by logunova-anastasia**

### CODE REVIEW

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -28,6 +30,13 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError
+
+        text_words = text.lower().replace('.', f' {self._end_of_word_token}').replace(
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:36]: 😲

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 1 above)
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: changed it

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -28,6 +32,19 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError
+
+        text_words = []
+        for i in text.lower().split():
+            if i[-1] in ('.', '!', '?'):
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:40]: why do you create the tuple on each iteration?

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -125,6 +204,11 @@ def get_conversion_generator_type(self, generator_type: int) -> str:  # type: ig
         Returns:
             (str): Name of the generator.
         """
+        if generator_type == 0:
+            return 'Greedy Generator'
+        if generator_type == 1:
+            return 'Top-P Generator'
+        return 'Beam Search Generator'
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:211]: what if `generator_type` equels 3/4/5/6?

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -220,6 +321,33 @@ def _calculate_perplexity(self, generated_text: str) -> float:  # type: ignore
                 or if methods used return None,
                 or if nothing was generated.
         """
+        if not isinstance(generated_text, str) or not generated_text:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:325]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -220,6 +321,33 @@ def _calculate_perplexity(self, generated_text: str) -> float:  # type: ignore
...
+
+        encoded = self._word_processor.encode(generated_text)
+
+        if not encoded:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:330]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 6]
```diff
@@ -220,6 +321,33 @@ def _calculate_perplexity(self, generated_text: str) -> float:  # type: ignore
...
+
+        n_gram_size = self._language_model.get_n_gram_size()
+        sum_log = 0.0
+
+        for index in range(n_gram_size - 1, len(encoded)):
+            context = tuple(encoded[index - n_gram_size + 1: index])
+            generated_tokens = self._language_model.generate_next_token(context)
+
+            if not generated_tokens:
+                raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:340]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 2 above)
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: changed it

File: lab_4_fill_words_by_ngrams/main.py [hunk 7]
```diff
@@ -39,6 +56,11 @@ def _put(self, element: str) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(element, str) or not element:
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py [hunk 8]
```diff
@@ -69,7 +105,7 @@ class TopPGenerator:
     """
 
     def __init__(
-        self, language_model: NGramLanguageModel, word_processor: WordProcessor, p_value: float
+            self, language_model: NGramLanguageModel, word_processor: WordProcessor, p_value: float
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py [hunk 9]
```diff
@@ -98,6 +137,43 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+        if not encoded:
+            raise ValueError
+
+        encoded_list = list(encoded)
+
+        for i in range(seq_len):
+            next_tokens = self._model.generate_next_token(tuple(encoded_list))
+
+            if next_tokens is None:
+                raise ValueError
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py [hunk 10]
```diff
@@ -98,6 +137,43 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
                 or if methods used return None.
         """
+        if not isinstance(seq_len, int) or not isinstance(
+                prompt, str) or seq_len <= 0:
+            raise ValueError
+
+        encoded = self._word_processor.encode(prompt)
+
+        if not encoded:
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py [hunk 11]
```diff
@@ -98,6 +137,43 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
                 or if sequence has inappropriate length,
                 or if methods used return None.
         """
+        if not isinstance(seq_len, int) or not isinstance(
+                prompt, str) or seq_len <= 0:
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 3 above)
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: changed it

File: lab_4_fill_words_by_ngrams/main.py (hunk 4 above)
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 5 above)
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 6 above)
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py [hunk 12]
```diff
@@ -98,6 +139,43 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+        if not encoded:
+            raise ValueError('TopPGenerator.run: encoded text is empty')
+
+        encoded_list = list(encoded)
+
+        for i in range(seq_len):
+            next_tokens = self._model.generate_next_token(tuple(encoded_list))
+
+            if next_tokens is None:
+                raise ValueError('TopPGenerator.run: generate_next_token returned None')
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:157]: You should see `TopPGenerator.run` on traceback. You can remove this prefic from messages

File: lab_4_fill_words_by_ngrams/main.py [hunk 13]
```diff
@@ -125,6 +206,9 @@ def get_conversion_generator_type(self, generator_type: int) -> str:  # type: ig
         Returns:
             (str): Name of the generator.
         """
+        types = {self.greedy: 'Greedy Generator', self.top_p: 'Top-P Generator',
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:209]: Can you move it to __init__?

File: lab_4_fill_words_by_ngrams/start.py [hunk 14]
```diff
@@ -10,8 +14,51 @@ def main() -> None:
     """
     with open("./assets/Harry_Potter.txt", "r", encoding="utf-8") as text_file:
         text = text_file.read()
-    result = None
+
+    processor = WordProcessor('.')
+    encoded = processor.encode(text)
+
+    if not(isinstance(encoded, tuple) and encoded):
+        return None
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/start.py:22]: raise?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 12 above)
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 13 above)
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/main.py]: yes, done!

File: lab_4_fill_words_by_ngrams/start.py (hunk 14 above)
Comment (logunova-anastasia) [reply lab_4_fill_words_by_ngrams/start.py]: changed

**Review by artyomtugaryov — APPROVED** [review]

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**